File: Python/Island Perimeter.py
from typing import List


class Solution:
    """
    Time:   O(n*m)
    Memory: O(1)
    """

    LAND = 1
    WATER = 0

    def islandPerimeter(self, grid: List[List[int]]) -> int:
        n, m = len(grid), len(grid[0])
        perimeter = 0

        for r in range(n):
            for c in range(m):
                if grid[r][c] != self.LAND:
                    continue
                if r == 0     or grid[r - 1][c] == self.WATER:
                    perimeter += 1
                if r == n - 1 or grid[r + 1][c] == self.WATER:
                    perimeter += 1
                if c == 0     or grid[r][c - 1] == self.WATER:
                    perimeter += 1
                if c + 1 == m or grid[r][c + 1] == self.WATER:
                    perimeter += 1

        return perimeter


class Solution:
    """
    Time:   O(n*m)
    Memory: O(1)
    """

    LAND = 1
    WATER = 0

    def islandPerimeter(self, grid: List[List[int]]) -> int:
        n, m = len(grid), len(grid[0])
        return sum(
            sum([
                r == 0     or grid[r - 1][c] == self.WATER,
                r == n - 1 or grid[r + 1][c] == self.WATER,
                c == 0     or grid[r][c - 1] == self.WATER,
                c + 1 == m or grid[r][c + 1] == self.WATER
            ]) for r in range(n)
            for c in range(m) if grid[r][c] == self.LAND
        )

File: Java/Island Perimeter.java
public class Solution {

    /**
     * Time:   O(n*m)
     * Memory: O(1)
     */
    public int islandPerimeter(int[][] grid) {
        int n = grid.length, m = grid[0].length;
        int perimeter = 0;

        for (int i = 0; i < n; ++i) {
            for (int j = 0; j < m; ++j) {
                if (grid[i][j] != LAND)
                    continue;

                if (i == 0 || grid[i - 1][j] == WATER)
                    perimeter += 1;
                if (i + 1 == n || grid[i + 1][j] == WATER)
                    perimeter += 1;
                if (j == 0 || grid[i][j - 1] == WATER)
                    perimeter += 1;
                if (j + 1 == m || grid[i][j + 1] == WATER)
                    perimeter += 1;
            }
        }
        return perimeter;
    }

    public static final int WATER = 0, LAND = 1;
}
